### PR TITLE
HLRC: Add ML delete filter action

### PR DESCRIPTION
**Pull request by iverase**
It adds delete ML filter action to the high level rest client.

Relates #29827

### CODE REVIEW

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java [hunk 1]
```diff
@@ -876,6 +877,16 @@ public void testFilterJob() throws Exception {
         MlFilter createdFilter = putFilterResponse.getResponse();
 
         assertThat(createdFilter, equalTo(mlFilter));
+
+        DeleteFilterRequest deleteFilterRequest = new DeleteFilterRequest(filterId);
```
Comment (benwtrent) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: please move this to its own test, almost all other of the CRUD operations are separated out to their own tests.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * A request to delete a machine learning filter
+ */
+public class DeleteFilterRequest implements Validatable {
+
+    private final String filter_id;
+
+    public DeleteFilterRequest(String filter_id) {
+        this.filter_id = Objects.requireNonNull(filter_id, "filter id is required");
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java]: message should match the parameter value `[filter_id] is required`.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MLRequestConverters.java [hunk 3]
```diff
@@ -476,4 +477,15 @@ static Request putFilter(PutFilterRequest putFilterRequest) throws IOException {
         request.setEntity(createEntity(putFilterRequest, REQUEST_BODY_CONTENT_TYPE));
         return request;
     }
+
+    static Request deleteFilter(DeleteFilterRequest deleteFilterRequest) throws IOException {
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MLRequestConverters.java]: the exception declaration is not necessary.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java (hunk 1 above)
Comment (benwtrent) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MachineLearningIT.java]: please move this to its own test, almost all other of the CRUD operations are separated out to their own tests.

File: docs/java-rest/high-level/ml/delete-filter.asciidoc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+[id="{upid}-{api}"]
+=== Delete Filter API
+Delete a {ml} filter.
+The API accepts a +{request}+ and responds
+with a +{response}+ object.
+
+[id="{upid}-{api}-request"]
+==== Delete Filter Request
+
+A `DeleteFilter` object requires a non-null `filterId`.
```
Comment (benwtrent) [inline docs/java-rest/high-level/ml/delete-filter.asciidoc]: This should be +{request}+ and not `DeleteFilter`

**Review by benwtrent — COMMENTED** [review]
Thanks a bunch for knocking this out :D. 

Just a couple of comments.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import org.elasticsearch.client.Validatable;
+
+import java.util.Objects;
+
+/**
+ * A request to delete a machine learning filter
+ */
+public class DeleteFilterRequest implements Validatable {
+
+    private final String filter_id;
```
Comment (benwtrent) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java]: Fields and parameters should be `camelCase`, so this should be `filterId`. 

Everything else looks good to me.

**Review by benwtrent — COMMENTED** [review]
Sorry for not spotting this before :(

File: client/rest-high-level/src/main/java/org/elasticsearch/client/MLRequestConverters.java [hunk 6]
```diff
@@ -476,4 +477,15 @@ static Request putFilter(PutFilterRequest putFilterRequest) throws IOException {
         request.setEntity(createEntity(putFilterRequest, REQUEST_BODY_CONTENT_TYPE));
         return request;
     }
+
+    static Request deleteFilter(DeleteFilterRequest deleteFilterRequest) {
+        String endpoint = new EndpointBuilder()
+            .addPathPartAsIs("_xpack")
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/MLRequestConverters.java]: `addPathPartAsIs("_xpack", "ml", "filters")`

File: client/rest-high-level/src/test/java/org/elasticsearch/client/MLRequestConvertersTests.java [hunk 7]
```diff
@@ -528,6 +529,16 @@ public void testPutFilter() throws IOException {
...
     }
 
+    public void testDeleteFilter() throws IOException {
+        MlFilter filter = MlFilterTests.createRandom("foo");
+        DeleteFilterRequest deleteFilterRequest = new DeleteFilterRequest(filter.getId());
+
+        Request request = MLRequestConverters.deleteFilter(deleteFilterRequest);
+
+        assertEquals(HttpDelete.METHOD_NAME, request.getMethod());
+        assertThat(request.getEndpoint(), equalTo("/_xpack/ml/filters/foo"));
```
Comment (hub-cap) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/MLRequestConvertersTests.java]: add `Assert.assertNull(request.getEntity());`

**Review by hub-cap — APPROVED** [review]
nits, ship it.

**Review by benwtrent — APPROVED** [review]

**Comment by iverase** [comment]
@elasticmachine test this please

**Comment by iverase** [comment]
@elasticmachine test this please

**Comment by iverase** [comment]
@elasticmachine test this please

**Comment by iverase** [comment]
@elasticmachine test this please

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * A request to delete a machine learning filter
+ */
+public class DeleteFilterRequest implements Validatable {
+
+    private final String filterId;
+
+    public DeleteFilterRequest(String filterId) {
+        this.filterId = Objects.requireNonNull(filterId, "[filter_id] is required");
```
Comment (hub-cap) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java]: should this still be `[filter_id]` ?

**Review by hub-cap — APPROVED** [review]
Hmm, 2 deletes, 2 diff responses in your PRs... I think we should consider the differences between the two of these and how to return a single, proper ACK. I wont let this hold up the review tho.

**Comment by benwtrent** [comment]
@hub-cap @iverase It does return a 404, 

https://github.com/elastic/elasticsearch/blob/9494e046e75d752aac26c66c01d090d3b2980991/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteFilterAction.java#L84-L86

File: client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java (hunk 8 above)
Comment (benwtrent) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/ml/DeleteFilterRequest.java]: @hub-cap yes, that holds with other ML nonNull checks. 

e.g.
https://github.com/elastic/elasticsearch/blob/fc774a377671d9f633513b3f8d954b237b377178/client/rest-high-level/src/main/java/org/elasticsearch/client/ml/job/config/AnalysisConfig.java#L341

**Comment by hub-cap** [comment]
Cool, so then keeping this one with no ignores=404 is a good thing. ty for verifying @benwtrent

**Comment by iverase** [comment]
@elasticmachine test this please

**Pull request merged**